### PR TITLE
Feature: improve logging & bugfixes

### PR DESCRIPTION
**Pull request by dgreene-r7**
This PR fixes a bug where mutations were occurring when the `storage.properties` object was being cloned by reference via the structured clone algorithm whenever a GET request went to the `/v1/conqueso/*` endpoint.

Second, it adds a check if the storage properties are rebuilding and if so, the `/v1/conqueso/*` endpoints return a `503` status code so that Archaius will use the previously stored value.

Finally, it adds logging around when updates that carry a change are assigned to `storage.properties` to hopefully track down bugs in software that consumes propsd.

### CODE REVIEW

File: lib/storage.js [hunk 1]
```diff
@@ -69,6 +72,13 @@ class Storage {
...
       merge(source.properties, properties);
     });
+
+    const diff = deepDiff(this.properties, properties);
+
+    if (diff) {
+      Log.info(`Calculated property diff`, {diff});
+    }
+
     this.properties = properties;
```
Comment (jeffreymyers-r7) [inline lib/storage.js]: Should this.properties only be updated if there's a diff?

File: lib/plugin-manager.js [hunk 2]
```diff
@@ -251,20 +251,20 @@ class PluginManager extends EventEmitter {
    */
   _registerSourceEvents(source) {
     source.on('startup', () => {
-      Log.verbose(`${source.name} started up.`);
+      Log.debug(`${source.name} started up.`);
```
Comment (jeffreymyers-r7) [inline lib/plugin-manager.js]: Consider logging startup and shutdown at an info level

File: lib/plugin-manager.js [hunk 3]
```diff
@@ -251,20 +251,20 @@ class PluginManager extends EventEmitter {
...
     });
 
     source.on('shutdown', () => {
-      Log.verbose(`${source.name} shut down.`);
+      Log.debug(`${source.name} shut down.`);
     });
 
     source.on('update', () => {
-      Log.verbose(`${source.name}'s data was updated from its underlying source data.`);
+      Log.debug(`${source.name}'s data was updated from its underlying source data.`);
```
Comment (jeffreymyers-r7) [inline lib/plugin-manager.js]: Consider logging this at an info level

**Comment by fmitchell-r7** [comment]
Thanks for tracking down this bug!

Is there any way we can write a test for this? It'd be nice to avoid updating `clone` or replacing it with another library in the future and reintroducing this issue.

File: lib/storage.js (hunk 1 above)
Comment (dgreene-r7) [inline lib/storage.js]: You're right but I'd like to understand the diffs before we confirm that properties should only be updated if there is a diff as calculated by this lib. I'm adding more diff logging so we can get some insight into the current behavior.

Comment (jeffreymyers-r7) [inline lib/storage.js]: Alright. I'm all for insight. :dollar:

**Comment by dgreene-r7** [comment]
> Is there any way we can write a test for this? It'd be nice to avoid updating clone or replacing it with another library in the future and reintroducing this issue.

I'm not entirely sure. Our best bet is probably to fix the library version or port their code into propsd. I'm amenable to either. Preferences?

File: lib/plugin-manager.js [hunk 4]
```diff
@@ -250,46 +250,55 @@ class PluginManager extends EventEmitter {
...
     this._ok = false;
 
+    const logMetadata = (source) ? {source: source.name, type: source.type} : {};
+
     // Only emit an error event if there are listeners.
     if (this.listeners('error').length > 0) {
-      this.emit('error', err);
+      this.emit('error', err, logMetadata);
+    } else {
+      Log.error(err, logMetadata);
```
Comment (jmanero-r7) [inline lib/plugin-manager.js]: This shouldn't be conditional

File: lib/source/consul.js [hunk 5]
```diff
@@ -343,19 +342,19 @@ class Consul extends EventEmitter {
...
 
-    Log.error(error, {
-      source: this.name,
-      type: this.type
-    });
-
     if (this.listeners('error').length > 0) {
-      this.emit('error', error);
+      this.emit('error', err);
+    } else {
```
Comment (jmanero-r7) [inline lib/source/consul.js]: This shouldn't be conditional

File: lib/control/v1/conqueso.js [hunk 6]
```diff
@@ -137,7 +116,7 @@ function Conqueso(app, storage) {
 
   route.get((req, res) => {
     res.set('Content-Type', 'text/plain');
-    res.status(HTTP_OK);
+    res.status(statusCode);
     res.end(makeJavaProperties(makeConquesoProperties(storage.properties)));
```
Comment (jeffreymyers-r7) [inline lib/control/v1/conqueso.js]: Can we avoid sending the body if statusCode != HTTP_OK?

**Comment by fmitchell-r7** [comment]
I'm trying to wrap my head around what's in the PR.

* clone function was replaced with [a library](https://github.com/pvorb/node-clone)
* INFO level logging around start, stop, update moved from sources to plugin manager
* verbose logging was added for deltas when plugins update
* 503 is returned (instead of 200) while properties are being computed and updated

@dgreene-r7, does that sum it up?

**Comment by dgreene-r7** [comment]
> I'm trying to wrap my head around what's in the PR.
> 
> clone function was replaced with a library
> INFO level logging around start, stop, update moved from sources to plugin manager
> verbose logging was added for deltas when plugins update
> 503 is returned (instead of 200) while properties are being computed and updated
> @dgreene-r7, does that sum it up?

Yup, that's everything. Normally I'd have split it out into multiple PRs but that urgency tho.

**Comment by rvanburen-r7** [comment]
I think this makes sense. though you probably want more eyes on this.

**Comment by jmanero-r7** [comment]
:shipit:

**Pull request merged**